storage/pebble/register_index.go

```diff
@@ -0,0 +1,172 @@
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage/pebble/registers"
+
+	"github.com/cockroachdb/pebble"
+)
+
+var (
+	// MinLookupKeyLen defines the minimum length for a valid lookup key
+	//
+	// Lookup keys use the following format:
+	//     [key] / [height]
+	// Where:
+	// - key: optional variable length field
+	// - height: 8 bytes representing the block height (uint64)
+	// - separator: '/' is used to separate variable length field
+	//
+	// Therefore the minimum key would be 1 byte + # of bytes for height
+	//     / [height]
+	MinLookupKeyLen = 1 + registers.HeightSuffixLen
+)
+
+type RegisterIndex struct {
+	store *Storage
+	owner flow.Address
+}
+
+// NewRegisters creates a new index instance at the provided height, all reads and
+// writes of the registers will happen at that height.
+// this is not concurrency safe.
+//
+// The register store does verify that the owner supplied is the one that was used before,
+// or that the heights are sequential.
+// This should be done by the caller.
+//
+// The RegisterIndex is modeled after `pebble.Registers` from `flow-go` but there are a few differences:
+//  1. The `flow-go` implementation creates its own independent batch when saving registers.
+//     The gateway needs to save the registers together with blocks and transaction so the batch
+//     is shared with that.
+//  2. The gateway does not need to store the owner address as all the registers are for the same owner.
+//  3. The gateway does not need pruning (yet) as the db is supposed to be much smaller.
+//  4. The owner and height checks are expected to be performed by the caller.
+func NewRegisters(
+	store *Storage,
+	owner flow.Address,
+) *RegisterIndex {
+	return &RegisterIndex{
+		store: store,
+		owner: owner,
+	}
+}
+
+// Get returns the register value for the given register ID at the given height.
+// Get will check that the owner is the same as the one used to create the index.
+func (r *RegisterIndex) Get(id flow.RegisterID, height uint64) (flow.RegisterValue, error) {
+	owner := flow.BytesToAddress([]byte(id.Owner))
+	if r.owner != flow.BytesToAddress([]byte(id.Owner)) {
+		return nil, registerOwnerMismatch(r.owner, owner)
+	}
+
+	lookupKey := newLookupKey(height, []byte(id.Key))
+	return r.lookupRegister(lookupKey.Bytes())
+}
+
+// Store stores the register entries for the given height to the given batch.
+// The batch does need to be indexed.
+//
+// Store will check that all the register entries are for the same owner.
+func (r *RegisterIndex) Store(entries flow.RegisterEntries, height uint64, batch *pebble.Batch) error {
+	for _, entry := range entries {
+		owner := flow.BytesToAddress([]byte(entry.Key.Owner))
+		if r.owner != owner {
+			return registerOwnerMismatch(r.owner, owner)
+		}
+
+		encoded := newLookupKey(height, []byte(entry.Key.Key)).Bytes()
+
+		err := batch.Set(encoded, entry.Value, nil)
+		if err != nil {
+			return fmt.Errorf("failed to set key: %w", err)
+		}
+	}
+
```

Comment:
Do we want to add a latest height key and update along with the updated registers? Otherwise, how do we know what is the latest height of the register store?

And to protect data integrity, it's better to ensure the height is the next height of the latest height.

Reply:
Since the pebble batch will also contain blocks, its much easier to only check the height integrity on the blocks.

blocks and registers will always be in sync since they will be batched together.